javascript/ql/src/semmle/javascript/security/performance/ReDoSUtil.qll

```diff
@@ -123,6 +123,23 @@ private class RegexpCharacterConstant extends RegExpConstant {
   RegexpCharacterConstant() { this.isCharacter() }
 }
 
+/**
+ * Holds if `term` is the chosen cannonical representative for all terms with string representation `str`.
+ *
+ * Using cannonical representatives gives a huge performance boost when working with tuples containing multiple `InputSymbol`s.
+ * The number of `InputSymbol`s is decreased by 3 orders of magnitude or more in some larger benchmarks.
+ */
+private predicate isCannonicalTerm(RegExpTerm term, string str) {
+  term =
+    rank[1](RegExpTerm t, Location loc, File file |
+      loc = t.getLocation() and
+      file = t.getFile() and
+      str = t.getRawValue()
```

Comment:
We would have to do that in two steps to prevent a blowup. So start with the current canonicalization, and then afterwards do your suggestion.  
And yes, I think we could possibly squeeze some more out of it, but I think we'll mostly get `\d` vs `[0-9]`. 

I'll give it a go.